// Copyright 2023 Ant Group Co., Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package secretflow.component;

import "secretflow/protos/component/comp.proto";

// Displays multiple read-only fields in groups.
message Descriptions {
  message Item {
    // Name of the field.
    string name = 1;

    string desc = 2;

    AttrType type = 3;

    Attribute value = 4;
  }

  // Name of the group.
  string name = 1;

  string desc = 2;

  repeated Item items = 3;
}

// Displays rows of data.
message Table {
  message HeaderItem {
    string name = 1;

    string desc = 2;

    AttrType type = 3;
  }

  message Row {
    string name = 1;

    string desc = 2;

    repeated Attribute items = 3;
  }

  // Name of the table
  string name = 1;

  string desc = 2;

  repeated HeaderItem headers = 3;

  repeated Row rows = 4;
}

// A division or a section of the page
message Div {
  message Child {
    // supported: descriptions, table, div
    string type = 1;

    Descriptions descriptions = 2;

    Table table = 3;

    Div div = 4;
  }

  // Name of the Div
  string name = 1;

  string desc = 2;

  repeated Child children = 3;
}

// A page of report.
message Tab {
  // Name of the tab.
  string name = 1;

  string desc = 2;

  repeated Div divs = 3;
}

message Report {
  // Name of the report.
  string name = 1;

  string desc = 2;

  repeated Tab tabs = 3;

  int32 err_code = 4;

  // err detail (json encoded message, structed error detail)
  string err_detail = 5;
}
